### association learning is used to discover interesting relationships or patterns in large dataset
### often applied to transactional data where items are bought or used together.
### works on the if and else statement such as if A the B. if element is called antecedent and then statement is called consedent.
### 3 techniques are used that are Support, Confidence and Lift
## 1) Support:
### the first step and is used to find the frequent bought items and is the straughtforwar calculation based on frequency.
### Support(A) = Transactions(A) / Total Transactions
## 2) Confidence:
### the second step that tells us how confident we can be that an item will be pruchased given that another item has been purchased.
### Conf (X==>Y) = P(Y|X) = (supp(X intersaction Y)) / (supp(X)) = (no of transactions containing X and Y) / (no of transactions containing X)
## Lift:
### lift (X==>Y) = (supp(X intersaction Y)) / (supp(X) * supp(Y))
### lift > 1: means that the two items are more likely to be bought together and hence keep both items together
### lift < 1: means that the two items are more likely to be bought separately and hence no need to keep both items together but shoould be placed close to each other.
### lift = 1: means that there is no association between the two items. 

### 3 types of association learning are as under
## 1) Apriori algo:
### Algorithm Steps:
#### Step 1 (k=1): Create a table (C1) with the support count of each item. Remove items with support counts below a minimum threshold.
#### Step 2 (k=2): Generate candidate sets (C2) by joining frequent itemsets from the previous step. Check subsets for frequency and prune accordingly.
#### Step 3 (k=3+): Continue generating candidate sets (C3, C4, etc.) and pruning based on the Apriori property.


In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [24]:
d_set=pd.read_csv("groceries - groceries.csv")
d_set.head()

Item(s)            Item 1               Item 2          Item 3  \
0        4      citrus fruit  semi-finished bread       margarine   
1        3    tropical fruit               yogurt          coffee   
2        1        whole milk                  NaN             NaN   
3        4         pip fruit               yogurt    cream cheese   
4        4  other vegetables           whole milk  condensed milk   

                     Item 4 Item 5 Item 6 Item 7 Item 8 Item 9  ... Item 23  \
0               ready soups    NaN    NaN    NaN    NaN    NaN  ...     NaN   
1                       NaN    NaN    NaN    NaN    NaN    NaN  ...     NaN   
2                       NaN    NaN    NaN    NaN    NaN    NaN  ...     NaN   
3              meat spreads    NaN    NaN    NaN    NaN    NaN  ...     NaN   
4  long life bakery product    NaN    NaN    NaN    NaN    NaN  ...     NaN   

  Item 24 Item 25 Item 26 Item 27 Item 28 Item 29 Item 30 Item 31 Item 32  
0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
2     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
3     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
4     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[5 rows x 33 columns]

### as it can be seen that there are many nan values and we need to remove them

### the type of the nan vaulues is always float and that of categroical data is a string and based on it we will separfate our data

In [25]:
# nan value type check
type(d_set['Item 5'][0])

float

In [26]:
# now we have to create a list of market in which all items other than nan will be kept
market=[]

# loop around the number of rows
for i in range(0, d_set.shape[0]): # .shape will run till the last row value

    # now we have to separate the nan value and keep all other values in the customer's list because those items are sold to them
    customers=[]

    # now loop around the columns to separate the nan type values
    for j in d_set.columns:

        # condition
        if type(d_set[j][i])==str:
            # append the values in customer's list
            customers.append(d_set[j][i])

    # finally append the customer data in te market list
    market.append(customers)
# the market list will contain all the nested list of the items that are bought by the customers
# since the market is a nested list and we need to convert it into a single numpy array as follows
l=[]

# loop around the market values
for i in market:
    # loop around the i
    for j in i:
        l.append(j)


In [27]:
# now count the frequency that will show the number of sold items in descending order
# import library
import collections

# count the items
counted=collections.Counter(l)
counted

Counter({'citrus fruit': 814,
         'semi-finished bread': 174,
         'margarine': 576,
         'ready soups': 18,
         'tropical fruit': 1032,
         'yogurt': 1372,
         'coffee': 571,
         'whole milk': 2513,
         'pip fruit': 744,
         'cream cheese': 390,
         'meat spreads': 42,
         'other vegetables': 1903,
         'condensed milk': 101,
         'long life bakery product': 368,
         'butter': 545,
         'rice': 75,
         'abrasive cleaner': 35,
         'rolls/buns': 1809,
         'UHT-milk': 329,
         'bottled beer': 792,
         'liquor (appetizer)': 78,
         'potted plants': 170,
         'cereals': 56,
         'white bread': 414,
         'bottled water': 1087,
         'chocolate': 488,
         'curd': 524,
         'flour': 171,
         'dishes': 173,
         'beef': 516,
         'frankfurter': 580,
         'soda': 1715,
         'chicken': 422,
         'sugar': 333,
         'fruit/vegetable juice': 711,
 

In [28]:
# convert the counted values into dataframe
# fisrt make a dictionary that will hold the item name and corresponding selling values
d={"Item name":counted.keys(), "values":counted.values()}

# now dataframe
df=pd.DataFrame(d).sort_values(by=['values'], ascending=False)
df.head()

Item name  values
7         whole milk    2513
11  other vegetables    1903
17        rolls/buns    1809
31              soda    1715
5             yogurt    1372

In [30]:
# now model
from mlxtend.preprocessing.transactionencoder import TransactionEncoder

# model
te=TransactionEncoder()

# fit the market data
te.fit(market)

# make dataframe of it
m_df=pd.DataFrame(te.transform(market), columns=te.columns_)
m_df.head()

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
0                  False     False             False             False   
1                  False     False             False             False   
2                  False     False             False             False   
3                  False     False             False             False   
4                  False     False             False             False   

   baby cosmetics  baby food   bags  baking powder  bathroom cleaner   beef  \
0           False      False  False          False             False  False   
1           False      False  False          False             False  False   
2           False      False  False          False             False  False   
3           False      False  False          False             False  False   
4           False      False  False          False             False  False   

   ...  turkey  vinegar  waffles  whipped/sour cream  whisky  white bread  \
0  ...   False    False    False               False   False        False   
1  ...   False    False    False               False   False        False   
2  ...   False    False    False               False   False        False   
3  ...   False    False    False               False   False        False   
4  ...   False    False    False               False   False        False   

   white wine  whole milk  yogurt  zwieback  
0       False       False   False     False  
1       False       False    True     False  
2       False        True   False     False  
3       False       False    True     False  
4       False        True   False     False  

[5 rows x 169 columns]

### result: above dataframe shows the relationship between two selling items. The False means these two items are not sold together

In [34]:
# now performing apriori algo
from mlxtend.frequent_patterns import apriori

# model
apriori(m_df, min_support=0.05, use_colnames=True, max_len=3).sort_values(by=['support'], ascending=False)

support                        itemsets
26  0.255516                    (whole milk)
15  0.193493              (other vegetables)
19  0.183935                    (rolls/buns)
23  0.174377                          (soda)
27  0.139502                        (yogurt)
2   0.110524                 (bottled water)
20  0.108998               (root vegetables)
24  0.104931                (tropical fruit)
22  0.098526                 (shopping bags)
21  0.093950                       (sausage)
16  0.088968                        (pastry)
6   0.082766                  (citrus fruit)
1   0.080529                  (bottled beer)
14  0.079817                    (newspapers)
5   0.077682                   (canned beer)
17  0.075648                     (pip fruit)
28  0.074835  (other vegetables, whole milk)
11  0.072293         (fruit/vegetable juice)
25  0.071683            (whipped/sour cream)
3   0.064870                   (brown bread)
9   0.063447                 (domestic eggs)
10  0.058973                   (frankfurter)
12  0.058566                     (margarine)
7   0.058058                        (coffee)
18  0.057651                          (pork)
29  0.056634        (rolls/buns, whole milk)
30  0.056024            (yogurt, whole milk)
4   0.055414                        (butter)
8   0.053279                          (curd)
0   0.052466                          (beef)
13  0.052364                       (napkins)

### result: from above values there are 3 combinations of 2 items that are sold together.
### but the support value for (other vegetables, whole milk) = 0.074835 is greater than theat of other two combinations. therefor, we will make use of (other vegetables, whole milk) for the next step